### PR TITLE
Ensure that MSBuildWorkspace deals with duplicate ProjectReferences

### LINKED ISSUE #31390: System.ArgumentException in SolutionState.CompilationTracker.cs
**Version Used**: Latest roslyn master sources

**Steps to Reproduce**:

1. Sync to latest roslyn master, assuming repo root here is `C:\roslyn`.
2. Build Roslyn.sln in Debug configuration
3. Build [AnalyzerRunner.csproj](https://github.com/dotnet/roslyn/blob/master/src/Tools/AnalyzerRunner/AnalyzerRunner.csproj)
4. Execute all IDE analyzers on Roslyn.sln using the following command from an admin command prompt: `C:\roslyn\Binaries\Debug\Exes\AnalyzerRunner\AnalyzerRunner.exe C:\roslyn\Binaries\Debug\Exes\AnalyzerRunner c:\roslyn\Roslyn.sln /concurrent`

**Expected Behavior**:
No exceptions

**Actual Behavior**:
`System.ArgumentException: An item with the same key has already been added.`
Also note that you might see assert from https://github.com/dotnet/roslyn/issues/31388

`.Add` invocation below seems to be throwing [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/SolutionState.CompilationTracker.cs,638) when processing `src\NuGet\VisualStudio\VS.ExternalAPIs.Roslyn.Package.csproj`:
```
                                var metadataReference = await solution.GetMetadataReferenceAsync(
                                    projectReference, this.ProjectState, cancellationToken).ConfigureAwait(false);
 
                                // A reference can fail to be created if a skeleton assembly could not be constructed.
                                if (metadataReference != null)
                                {
                                    newReferences.Add(metadataReference);
                                    metadataReferenceToProjectId.Add(metadataReference, projectReference.ProjectId);
                                }
                                else
                                {
                                    hasSuccessfullyLoaded = false;
                                }
```

**Call stack:**
```
Failed to analyze VS.ExternalAPIs.Roslyn.Package with System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<FinalizeCompilationAsync>d__30.MoveNext() in C:\roslyn_2\src\Workspaces\Core\Portable\Workspace\Solution\SolutionState.CompilationTracker.cs:line 638
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<BuildCompilationInfoFromScratchAsync>d__24.MoveNext() in C:\roslyn_2\src\Workspaces\Core\Portable\Workspace\Solution\SolutionState.CompilationTracker.cs:line 478
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<GetOrBuildCompilationInfoAsync>d__22.MoveNext() in C:\roslyn_2\src\Workspaces\Core\Portable\Workspace\Solution\SolutionState.CompilationTracker.cs:line 407
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.SolutionState.CompilationTracker.<GetCompilationSlowAsync>d__19.MoveNext() in C:\roslyn_2\src\Workspaces\Core\Portable\Workspace\Solution\SolutionState.CompilationTracker.cs:line 316
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at AnalyzerRunner.Program.<GetProjectAnalysisResultAsync>d__10.MoveNext() in C:\roslyn_2\src\Tools\AnalyzerRunner\Program.cs:line 437
```

### PR DESCRIPTION
**Pull request by jasonmalinowski**
The workspace isn't supposed to allow duplicate ProjectReferences, but currently due to #12101 it does. This dedups (correctly) at the surface and also adds a test to ensure this isn't broken.

Fixes #31390. Further work will address #12101, but no matter how we address that it's good to have tests at the MSBuildWorkspace layer to make sure some additional bugs don't creep in there in it's processing.

Special thanks to @alexbuyn for #32557 that helped alert to the root cause here, and @aolszowka for also chiming in with a bit of details about their scenario to confirm this should help fix it.

<details><summary>Ask Mode template</summary>

### Customer scenario

Customer uses a tool built atop MSBuildWorkspace, and they have a project with duplicate ProjectReferences where the exact same project is being referenced twice in the exact same way. This tool will probably crash when ran on this solution.

### Bugs this fixes

#31390

### Workarounds, if any

Fix the duplicate project reference. Unfortunately the crash is deep within the CompilationTracker, which won't really hint to the user that anything is wrong.

### Risk

Very low, the actual core fix is just replacing an ImmutableArray with an ImmutableHashSet. This API is not used in Visual Studio and thus does not pose a risk there.

### Performance impact

None.

### Is this a regression from a previous update?

Not clear. The underlying Workspace bug is positively ancient.

### Root cause analysis

We didn't have any tests, and it wasn't obvious the underlying Workspace bug would actual result in crashes.

### How was the bug found?

Customer reports.

</details>

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
While doing this change I noticed that the files that MSBuildWorkspace tests consumed were in a project that wasn't MSBuildWorkspaceTests...except it was only being (effectively) used in one place. So the first two commits in this PR are me cleaning that up. The last one is the real fix.

**Comment by aolszowka** [comment]
I just want to chime in again and say that I completely agree that this should fall over when it finds a duplicate; this is bad data in our mind and we want to remove the duplicate and then ideally root cause how it got dupe'd. Good exception messages make this possible.

In our case, for this particular project that triggered the error:

1. Had a ProjectReference to the Project and an AssemblyReference to the File
2. We made a push a few weeks ago to convert all of our AssemblyReferences to ProjectReferences using some in house tooling (Which we opened sourced here: https://github.com/aolszowka/MsBuildConvertToProjectReference)
3. This converter did not take into account checking for duplicates  (I have opened up https://github.com/aolszowka/MsBuildConvertToProjectReference/issues/3 to fix this in our tooling)

As I note in the above Visual Studio did not seem to care; nor did MSBuild Proper when building; it was only the StyleCop Analyzer that kicked out. We have other tooling that uses this (basically one to perform formatting of our code base based on some EditorConfig settings) however we do not utilize the Projects as written we load up all of those files into an adhoc solution.

Thank  you for the fix.

File: src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs [hunk 1]
```diff
@@ -17,10 +17,10 @@ private partial class Worker
         {
             private readonly struct ResolvedReferences
             {
-                public ImmutableArray<ProjectReference> ProjectReferences { get; }
+                public ImmutableHashSet<ProjectReference> ProjectReferences { get; }
```
Comment (DustinCampbell) [inline src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs]: Are we concerned about ordering being blown out here?

Comment (jasonmalinowski) [reply src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs]: @DustinCampbell I wasn't concerned; I don't think reference ordering has any semantic meaning, and also wasn't sure if MSBuild and everything else is maintaining ordering anyways. Looking at the Workspace code, we do maintain the ordering of ProjectReferences, but give you no way to specify the full ordering of the references that go into the Compilation: we always add metadata references before project references.

Comment (jasonmalinowski) [reply src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs]: Per @jaredpar he doesn't think ordering matters.

Comment (jaredpar) [reply src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs]: I said I didn't think ordering matters. I didn't say we should make business decisions on my opinion 😉

**Pull request merged**